File: demo/backend/middlewares/logging.js
const { v4: uuidv4 } = require("uuid");

const logReqResp = () => (req, res, next) => {
  const start = Date.now();
  req.requestId = req.headers["request-id"] || uuidv4();
  const { method, url } = req;

  console.log(
    `Request request-id=${req.requestId}: [${new Date().toISOString()}] ${method} ${url}`
  );

  res.on("finish", () => {
    const duration = Date.now() - start;
    console.log(
      `Response request-id=${req.requestId}: [${new Date().toISOString()}] ${method} ${url} - ${res.statusCode} (${duration}ms)`
    );
  });

  next();
};

module.exports = {
    logReqResp
};

File: demo/backend/middlewares/logging.py
import uuid
from flask import request

def log_req_resp(app):
    @app.before_request
    def log_request():
        request_id = request.headers.get("request-id", str(uuid.uuid4()))
        request.environ["request_id"] = request_id
        app.logger.info(f"Request ID={request_id}: {request.method} {request.path}")

    @app.after_request
    def log_response(response):
        request_id = request.environ.get("request_id")
        app.logger.info(f"Response ID={request_id}: {response.status}")
        return response
